src/constructors/adjacency.c

```diff
@@ -858,3 +900,609 @@ igraph_error_t igraph_adjlist(igraph_t *graph, const igraph_adjlist_t *adjlist,
+                IGRAPH_EINVAL);
+    }
+    for (igraph_integer_t to = 0; p[to] < no_of_edges; to++) {
+        for (; from < p[to + 1]; from++) {
+            if (to >= i[from]) {
+                igraph_integer_t multi = entry[from];
+                if (to == i[from]) {
+                    IGRAPH_CHECK(igraph_i_adjust_loop_edge_count(&multi, loops));
+                }
+                for (igraph_integer_t count = 0; count < multi; count++) {
+                    VECTOR(*edges)[e++] = i[from];
+                    VECTOR(*edges)[e++] = to;
+                }
+            }
+        }
+    }
+    igraph_vector_int_resize(edges, e);
+
+    return IGRAPH_SUCCESS;
+}
+
+/**
+ * \ingroup generators
+ * \function igraph_sparse_adjacency
+ * \brief Creates a graph from a sparse adjacency matrix.
+ *
+ * This has the same functionality as \ref igraph_adjacency(), but uses
+ * a column-compressed adjacency matrix.
+ *
+ * Time complexity: O(|E|),
+ * where |E| is the number of edges in the graph.
+ */
+
+igraph_error_t igraph_sparse_adjacency(igraph_t *graph, igraph_sparsemat_t *adjmatrix,
+        igraph_adjacency_t mode, igraph_loops_t loops) {
+
+    igraph_vector_int_t edges = IGRAPH_VECTOR_NULL;
+    CS_INT no_of_nodes = adjmatrix->cs->m;
+    CS_INT no_of_edges = adjmatrix->cs->p[adjmatrix->cs->n];
+    if (no_of_edges) {
+        no_of_edges *= igraph_sparsemat_max(adjmatrix);
+    }
+
+    if (!igraph_sparsemat_is_cc(adjmatrix)) {
+        IGRAPH_ERROR("Sparse adjacency matrix should be in column-compressed "
+               "form.", IGRAPH_EINVAL);
+    }
+    if (no_of_nodes != adjmatrix->cs->n) {
+        IGRAPH_ERROR("Adjacency matrix is non-square.", IGRAPH_NONSQUARE);
+    }
+
+    if (no_of_nodes != 0 && igraph_sparsemat_min(adjmatrix) < 0) {
+        IGRAPH_ERRORF("Edge counts should be non-negative, found %g.", IGRAPH_EINVAL,
+                igraph_sparsemat_min(adjmatrix));
+    }
+
+    IGRAPH_VECTOR_INT_INIT_FINALLY(&edges, no_of_edges * 2);
+
+    /* Collect the edges */
+    switch (mode) {
+    case IGRAPH_ADJ_DIRECTED:
+        IGRAPH_CHECK(igraph_i_sparse_adjacency_directed(adjmatrix, &edges, loops));
+        break;
+    case IGRAPH_ADJ_MAX:
+        IGRAPH_CHECK(igraph_i_sparse_adjacency_max(adjmatrix, &edges, loops));
+        break;
+    case IGRAPH_ADJ_UNDIRECTED:
+        IGRAPH_CHECK(igraph_i_sparse_adjacency_undirected(adjmatrix, &edges, loops));
+        break;
+    case IGRAPH_ADJ_UPPER:
+        IGRAPH_CHECK(igraph_i_sparse_adjacency_upper(adjmatrix, &edges, loops));
+        break;
+    case IGRAPH_ADJ_LOWER:
+        IGRAPH_CHECK(igraph_i_sparse_adjacency_lower(adjmatrix, &edges, loops));
+        break;
+    case IGRAPH_ADJ_MIN:
+        IGRAPH_CHECK(igraph_i_sparse_adjacency_min(adjmatrix, &edges, loops));
+        break;
+    case IGRAPH_ADJ_PLUS:
+        IGRAPH_CHECK(igraph_i_sparse_adjacency_directed(adjmatrix, &edges, loops));
```

Comment:
This should be `igraph_i_sparse_adjacency_plus()` -- but since this issue was not caught by the unit tests, I wonder whether this indicates that the tests are not thorough enough (or that in the sparse case the `_plus()` variant is actually not needed at all?).

Reply:
I think it was completely by accident, but yes they should be equal. I've removed _plus for the unweighted version.